azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/compute/extensions/AzureComputeImageExtension.java

```diff
@@ -0,0 +1,135 @@
+                              @Named(Constants.PROPERTY_USER_THREADS) ListeningExecutorService userExecutor,
+                              VMImageToImage imageReferenceToImage) {
+      this.userExecutor = userExecutor;
+      this.group = azureComputeConstants.azureResourceGroup();
+      this.imageReferenceToImage = imageReferenceToImage;
+      this.api = api;
+   }
+
+   @Override
+   public ImageTemplate buildImageTemplateFromNode(String name, String id) {
+      name = name.toLowerCase();
+      return new ImageTemplateBuilder.CloneImageTemplateBuilder().nodeId(id).name(name).build();
+   }
+
+   @Override
+   public ListenableFuture<Image> createImage(ImageTemplate template) {
+      final CloneImageTemplate cloneTemplate = (CloneImageTemplate) template;
+      String id = cloneTemplate.getSourceNodeId();
+      final String storageAccountName = id.replaceAll("[^A-Za-z0-9 ]", "") + "stor";
+
+      boolean generalized = false;
+      while (!generalized) {
+         try {
+            api.getVirtualMachineApi(group).generalize(id);
+         } finally {
+            generalized = true;
+         }
+      }
+
+      final String[] disks = new String[2];
+      URI uri = api.getVirtualMachineApi(group).capture(id, cloneTemplate.getName(), CONTAINER_NAME);
+      if (uri != null){
+         boolean jobDone = Predicates2.retry(new Predicate<URI>() {
+            @Override public boolean apply(URI uri) {
+               try {
+                  List<ResourceDefinition> definitions = api.getJobApi().captureStatus(uri);
+                  if (definitions != null) {
+                     for (ResourceDefinition definition : definitions) {
+                        LinkedTreeMap<String, String> properties = (LinkedTreeMap<String, String>) definition.properties();
+                        Object storageObject = properties.get("storageProfile");
+                        LinkedTreeMap<String, String> properties2 = (LinkedTreeMap<String, String>) storageObject;
+                        Object osDiskObject = properties2.get("osDisk");
+                        LinkedTreeMap<String, String> osProperties = (LinkedTreeMap<String, String>) osDiskObject;
+                        Object dataDisksObject = properties2.get("dataDisks");
+                        ArrayList<Object> dataProperties = (ArrayList<Object>) dataDisksObject;
+                        LinkedTreeMap<String, String> datadiskObject = (LinkedTreeMap<String, String>) dataProperties.get(0);
+
+                        disks[0] = osProperties.get("name");
+                        disks[1] = datadiskObject.get("name");
+
+                     }
+                     return true;
+                  }
+                  return false;
+               }
+               catch (Exception e) {
+                  return false;
+               }
+            }
+         }, 15 * 1000 /* 15 second timeout */).apply(uri);
+
+      }
```

Comment:
With this approach the method is blocking all the time and returning a future that returns immediately. Instead, the heavy logic should be done by the future to avoid blocking users with this potentially expensive operation.